From 4c257bf8e76d8ebb34749d639354f5aa06d5f3d6 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <adisi@codeaurora.org>
Date: Mon, 13 Sep 2021 09:32:32 +0530
Subject: [PATCH] ath12k: fix workqueue not getting destroyed after rmmod

This change is to destroy the workqueues after the modules
are removed from kernel. Earlier, the workqueues were
not destroyed and left as it is.

If workqueues are not destroyed, it could lead to kernel
memory scarcity in a longer run. This could affect self and
other drivers workability as well.

Signed-off-by: Aditya Kumar Singh <adisi@codeaurora.org>
Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 46f6eab..3fac559 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1374,7 +1374,10 @@ void ath12k_core_deinit(struct ath12k_base *ab)
 void ath12k_core_free(struct ath12k_base *ab)
 {
 	del_timer_sync(&ab->rx_replenish_retry);
+	flush_workqueue(ab->workqueue_aux);
 	destroy_workqueue(ab->workqueue_aux);
+	
+	flush_workqueue(ab->workqueue);
 	destroy_workqueue(ab->workqueue);
 	kfree(ab);
 }
-- 
2.17.1

